File: LWThemeManager/Classes/LWThemeManager.m
//
//  LWThemeManager.m
//  MyInputMethod
//
//  Created by luowei on 15/8/18.
//  Copyright (c) 2015年 luowei. All rights reserved.
//

#import "LWThemeManager.h"

static NSString *const Dir_Themes = @"themes";

@implementation LWThemeManager{
    NSMutableDictionary *_theme;
    NSString *_currentName;
}

+ (LWThemeManager *)sharedInstance {
    __strong static LWThemeManager *sharedInstance;

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[LWThemeManager alloc] init];
    });

    return sharedInstance;
}

+(NSString *)pathInBundleWithFileName:(NSString *)fileName {
    return [LWThemeManager bundlePathNamed:fileName ofBundle:@"KeyboardTheme.bundle"];
}

+(NSString*)bundlePathNamed:(NSString*)name ofBundle:(NSString*)bundleName{
    NSString *file_name = [NSString stringWithFormat:@"%@",name];
    NSString *resourcePath = [[NSBundle mainBundle]resourcePath];
    NSString *bundlePath = [resourcePath stringByAppendingPathComponent:bundleName];
    NSString *file_path = [bundlePath stringByAppendingPathComponent:file_name];
    
    return file_path;
}

-(NSString *)currentName {
    return _currentName;
}

-(NSMutableDictionary *)theme {
    if(_currentName && _theme){
        return _theme;
    }
    [self updateThemeWithName:_currentName ?: @"default"];
    return _theme;
}

//把theme数据保存到文件
-(void)setThemeValue:(id)value forKey:(NSString *)key{
    [[LWThemeManager sharedInstance].theme setValue:value forKey:key];

    //把theme数据写入到docment的文件中
    NSString *plistName = [NSString stringWithFormat:@"%@Theme.plist", _currentName];

    NSString *documentsPath = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *themePlistPath = [NSString stringWithFormat:@"%@/%@/%@/%@", documentsPath, Dir_Themes, _currentName, plistName];
    [self.theme writeToFile:themePlistPath atomically:YES];
}

//从当前主题新建一个主题
-(BOOL)copyANewThemeWithName:(NSString *)name {
    //判断当前Plist是否存在,不存在用default plist
    NSString *plistName = [NSString stringWithFormat:@"%@Theme.plist",_currentName ?: @"default"];
    NSString *documentsPath = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *currentPlistPath = [NSString stringWithFormat:@"%@/%@/%@/%@", documentsPath, Dir_Themes, _currentName, plistName];

    NSFileManager *fmanager = [NSFileManager defaultManager];
    if(![fmanager fileExistsAtPath:currentPlistPath]){
        currentPlistPath = [LWThemeManager pathInBundleWithFileName:plistName];
    }

    //创建name子目录
    NSError *error;
    NSString *nameDirPath = [NSString stringWithFormat:@"%@/%@/%@", documentsPath, Dir_Themes, name];
    if(![fmanager fileExistsAtPath:nameDirPath]){
        [fmanager createDirectoryAtPath:nameDirPath withIntermediateDirectories:YES attributes:nil error:&error];
    }

    //拷贝
    NSString *nameFilePath = [NSString stringWithFormat:@"%@/%@Theme.plist", nameDirPath, name];
    [fmanager copyItemAtPath:currentPlistPath toPath:nameFilePath error:&error]; //拷贝
    if(error){
        return NO;
    }
    return YES;
}

//恢复默认主题设置
-(void)recoverDefaultTheme{
    //NSString *bundleFilePath = [[NSBundle mainBundle] pathForResource:@"defaultTheme" ofType:@"plist"];
    [self updateThemeWithName:@"default"];
}

//根据name更新主题
-(void)updateThemeWithName:(NSString *)name {
    //若theme设置没变化，且theme字典不为空
    if([name isEqualToString:_currentName] && _theme!=nil ){
        return;
    }

    //拼接path
    NSString *plistName = [NSString stringWithFormat:@"%@Theme.plist",name ?: @"default"];
    NSString *documentsPath = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *nameDirPath = [NSString stringWithFormat:@"%@/%@/%@", documentsPath, Dir_Themes, name];
    NSString *docPlistPath = [NSString stringWithFormat:@"%@/%@", nameDirPath, plistName];
    NSString *bundPlistPath = [LWThemeManager pathInBundleWithFileName:plistName];
    NSString *defaultPlistPath = [LWThemeManager pathInBundleWithFileName:@"defaultTheme.plist"];

    //判断是否存在
    NSFileManager *fmanager = [NSFileManager defaultManager];
    if(![fmanager fileExistsAtPath:docPlistPath]){  //document下不存在

        NSError *error;
        //创建name子目录
        if(![fmanager fileExistsAtPath:nameDirPath]){
            [fmanager createDirectoryAtPath:nameDirPath withIntermediateDirectories:YES attributes:nil error:&error];
        }

        //拷贝,把 Bundle里下的文件拷贝到 name目录下
        if([fmanager fileExistsAtPath:docPlistPath]){  //如果已经存在了
            //[NSDictionary dictionaryWithContentsOfURL:[NSURL fileURLWithPath:plistPath] error:nil];
            _theme = [NSDictionary dictionaryWithContentsOfFile:docPlistPath].mutableCopy;
        }else{
            //判断bundPlistPath是否存在
            if([fmanager fileExistsAtPath:bundPlistPath]){
                [fmanager copyItemAtPath:bundPlistPath toPath:docPlistPath error:&error];
                if (error) {
                    _theme = [NSDictionary dictionaryWithContentsOfFile:bundPlistPath].mutableCopy;
                }else{
                    _theme = [NSDictionary dictionaryWithContentsOfFile:docPlistPath ].mutableCopy;
                }

            }else{
                _theme = [NSDictionary dictionaryWithContentsOfFile:defaultPlistPath].mutableCopy;
            }

        }

    }else{
        _theme = [NSDictionary dictionaryWithContentsOfFile:docPlistPath ].mutableCopy;
    }

    NSAssert(_theme!=nil, @"Error: 主题皮肤配置不能为空");

    _currentName=name;
}


@end

File: LWThemeManager_swift/Classes/LWThemeManager.swift
//
//  LWThemeManager.swift
//  LWThemeManager
//
//  Swift version of theme manager
//  Copyright (c) 2015 luowei. All rights reserved.
//

import UIKit

// MARK: - Constants
public let KeyTheme = "Key_Theme"

// MARK: - Theme Value Helpers
public extension LWThemeManager {

    /// Get float value from theme by key
    func floatValue(forKey key: String) -> Float {
        return (theme[key] as? NSNumber)?.floatValue ?? 0.0
    }

    /// Get string value from theme by key
    func stringValue(forKey key: String) -> String? {
        return theme[key] as? String
    }

    /// Get UIColor value from theme by key
    func colorValue(forKey key: String) -> UIColor? {
        guard let rgbaString = theme[key] as? String else { return nil }
        return UIColor.theme_color(withRGBAString: rgbaString)
    }

    /// Get CGColor value from theme by key
    func cgColorValue(forKey key: String) -> CGColor? {
        return colorValue(forKey: key)?.cgColor
    }

    /// Set float value to theme
    func setFloatValue(_ value: Float, forKey key: String) {
        setThemeValue(NSNumber(value: value), forKey: key)
    }

    /// Set color value to theme
    func setColorValue(_ color: UIColor, forKey key: String) {
        if let rgbaString = color.theme_rgbaString() {
            setThemeValue(rgbaString, forKey: key)
        }
    }

    /// Set CGColor value to theme
    func setCGColorValue(_ color: CGColor, forKey key: String) {
        if let rgbaString = UIColor(cgColor: color).theme_rgbaString() {
            setThemeValue(rgbaString, forKey: key)
        }
    }
}

// MARK: - LWThemeManager
public class LWThemeManager {

    // MARK: - Singleton
    public static let shared = LWThemeManager()

    // MARK: - Properties
    private var _theme: [String: Any]?
    private var _currentName: String?

    private static let themesDirectory = "themes"

    private init() {}

    // MARK: - Public Methods

    /// Get path in bundle with file name
    /// - Parameter fileName: The file name
    /// - Returns: Full path in bundle
    public static func pathInBundle(withFileName fileName: String) -> String {
        return bundlePath(named: fileName, ofBundle: "KeyboardTheme.bundle")
    }

    /// Get current theme name
    public var currentName: String {
        return _currentName ?? "default"
    }

    /// Get current theme dictionary
    public var theme: [String: Any] {
        if let currentName = _currentName, let theme = _theme {
            return theme
        }
        updateTheme(withName: _currentName ?? "default")
        return _theme ?? [:]
    }

    /// Set theme value for key and save to file
    /// - Parameters:
    ///   - value: The value to set
    ///   - key: The key
    public func setThemeValue(_ value: Any, forKey key: String) {
        var currentTheme = theme
        currentTheme[key] = value
        _theme = currentTheme

        // Write theme data to document directory
        let plistName = "\(_currentName ?? "default")Theme.plist"
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let themePlistPath = "\(documentsPath)/\(Self.themesDirectory)/\(_currentName ?? "default")/\(plistName)"

        let dict = currentTheme as NSDictionary
        dict.write(toFile: themePlistPath, atomically: true)
    }

    /// Update specific theme file's key-value
    /// - Parameters:
    ///   - name: Theme name (nil to use current theme)
    ///   - value: The value to set
    ///   - key: The key
    public func updateTheme(withName name: String?, value: Any, forKey key: String) {
        let themeName = name ?? _currentName ?? "default"
        let plistName = "\(themeName)Theme.plist"

        // Check if file exists
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let themePlistPath = "\(documentsPath)/\(Self.themesDirectory)/\(themeName)/\(plistName)"

        guard FileManager.default.fileExists(atPath: themePlistPath) else {
            return
        }

        // Load, update and save
        if var theme = NSDictionary(contentsOfFile: themePlistPath) as? [String: Any] {
            theme[key] = value
            let dict = theme as NSDictionary
            dict.write(toFile: themePlistPath, atomically: true)
        }
    }

    /// Remove theme by name
    /// - Parameter themeName: The theme name to remove
    public func removeTheme(withName themeName: String?) {
        guard let themeName = themeName else { return }

        let plistName = "\(themeName)Theme.plist"
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let themePlistPath = "\(documentsPath)/\(Self.themesDirectory)/\(themeName)/\(plistName)"

        guard FileManager.default.fileExists(atPath: themePlistPath) else {
            return
        }

        // Delete
        try? FileManager.default.removeItem(atPath: themePlistPath)
    }

    /// Copy a new theme from current theme
    /// - Parameter name: New theme name
    /// - Returns: Success or failure
    @discardableResult
    public func copyNewTheme(withName name: String) -> Bool {
        // Check if current plist exists, use default if not
        let plistName = "\(_currentName ?? "default")Theme.plist"
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        var currentPlistPath = "\(documentsPath)/\(Self.themesDirectory)/\(_currentName ?? "default")/\(plistName)"

        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: currentPlistPath) {
            currentPlistPath = Self.pathInBundle(withFileName: plistName)
        }

        // Create name subdirectory
        let nameDirPath = "\(documentsPath)/\(Self.themesDirectory)/\(name)"
        if !fileManager.fileExists(atPath: nameDirPath) {
            try? fileManager.createDirectory(atPath: nameDirPath, withIntermediateDirectories: true)
        }

        // Copy
        let nameFilePath = "\(nameDirPath)/\(name)Theme.plist"
        do {
            try fileManager.copyItem(atPath: currentPlistPath, toPath: nameFilePath)
            return true
        } catch {
            return false
        }
    }

    /// Recover default theme settings
    public func recoverDefaultTheme() {
        let plistName = "\(_currentName ?? "default")Theme.plist"
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let nameDirPath = "\(documentsPath)/\(Self.themesDirectory)/\(_currentName ?? "default")"
        let docPlistPath = "\(nameDirPath)/\(plistName)"
        let bundlePlistPath = Self.pathInBundle(withFileName: plistName)

        let fileManager = FileManager.default

        // Remove existing if exists
        if fileManager.fileExists(atPath: docPlistPath) {
            try? fileManager.removeItem(atPath: docPlistPath)
        }

        // Copy from bundle
        do {
            try fileManager.copyItem(atPath: bundlePlistPath, toPath: docPlistPath)
            _theme = NSDictionary(contentsOfFile: docPlistPath) as? [String: Any]
        } catch {
            _theme = NSDictionary(contentsOfFile: bundlePlistPath) as? [String: Any]
        }
    }

    /// Update theme with name
    /// - Parameter name: Theme name
    public func updateTheme(withName name: String) {
        // If theme setting hasn't changed and theme dictionary is not empty
        if name == _currentName && _theme != nil {
            return
        }

        // Build paths
        let themeName = name.isEmpty ? "default" : name
        let plistName = "\(themeName)Theme.plist"
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let nameDirPath = "\(documentsPath)/\(Self.themesDirectory)/\(themeName)"
        let docPlistPath = "\(nameDirPath)/\(plistName)"
        let bundlePlistPath = Self.pathInBundle(withFileName: plistName)
        let defaultPlistPath = Self.pathInBundle(withFileName: "defaultTheme.plist")

        let fileManager = FileManager.default

        // Check if exists in documents
        if !fileManager.fileExists(atPath: docPlistPath) {
            // Create subdirectory
            if !fileManager.fileExists(atPath: nameDirPath) {
                try? fileManager.createDirectory(atPath: nameDirPath, withIntermediateDirectories: true)
            }

            // Copy bundle file to documents
            if fileManager.fileExists(atPath: docPlistPath) {
                _theme = NSDictionary(contentsOfFile: docPlistPath) as? [String: Any]
            } else {
                // Check if bundle plist exists
                if fileManager.fileExists(atPath: bundlePlistPath) {
                    do {
                        try fileManager.copyItem(atPath: bundlePlistPath, toPath: docPlistPath)
                        _theme = NSDictionary(contentsOfFile: docPlistPath) as? [String: Any]
                    } catch {
                        _theme = NSDictionary(contentsOfFile: bundlePlistPath) as? [String: Any]
                    }
                } else {
                    _theme = NSDictionary(contentsOfFile: defaultPlistPath) as? [String: Any]
                }
            }
        } else {
            _theme = NSDictionary(contentsOfFile: docPlistPath) as? [String: Any]
        }

        assert(_theme != nil, "Error: Theme configuration cannot be empty")

        _currentName = themeName
    }

    // MARK: - Private Methods

    private static func bundlePath(named name: String, ofBundle bundleName: String) -> String {
        let resourcePath = Bundle.main.resourcePath ?? ""
        let bundlePath = (resourcePath as NSString).appendingPathComponent(bundleName)
        let filePath = (bundlePath as NSString).appendingPathComponent(name)
        return filePath
    }
}
